src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -7177,169 +7206,274 @@ internal static List<CompletionResult> CompleteStatementFlags(TokenKind kind, st
             return results;
         }
 
+        private static PSTypeName GetNestedHashtableKeyType(TypeInferenceContext typeContext, PSTypeName parentType, IList<string> nestedKeys)
+        {
+            var currentType = parentType;
+            // The nestedKeys list should have the outer most key as the last element, and the inner most key as the first element
+            // If we fail to resolve the type of any key we return null
+            for (int i = nestedKeys.Count - 1; i >= 0; i--)
+            {
+                if (currentType is null)
+                {
+                    return null;
+                }
+
+                var typeMembers = typeContext.GetMembersByInferredType(currentType, false, null);
+                currentType = null;
+                foreach (var member in typeMembers)
+                {
+                    if (member is PropertyInfo propertyInfo)
+                    {
+                        if (propertyInfo.Name.Equals(nestedKeys[i], StringComparison.OrdinalIgnoreCase))
+                        {
+                            currentType = new PSTypeName(propertyInfo.PropertyType);
+                            break;
+                        }
+                    }
+                    else if (member is PropertyMemberAst memberAst && memberAst.Name.Equals(nestedKeys[i], StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (memberAst.PropertyType is null)
+                        {
+                            return null;
+                        }
+                        else
+                        {
+                            currentType = new PSTypeName(memberAst.PropertyType.TypeName);
+                        }
+                        break;
+                    }
+                }
+            }
+
+            return currentType;
+        }
+
         internal static List<CompletionResult> CompleteHashtableKey(CompletionContext completionContext, HashtableAst hashtableAst)
         {
+            Ast previousAst = hashtableAst;
+            Ast parentAst = hashtableAst.Parent;
+            string parameterName = null;
+            var nestedHashtableKeys = new List<string>();
+
+            // This loop determines if it's a nested hashtable and what the outermost hashtable is used for (Dynamic keyword, command argument, etc.)
+            while (parentAst is not null)
+            {
+                if (parentAst is HashtableAst parentTable)
+                {
```

Comment:
I think you may want to check if `previousAst is HashtableAst` before running the `foreach` loop. Consider this scenario: `xxx = @{ abc = @(@{d<tab>}) }`

Reply:
I am checking it before the foreach loop. Either I'm misunderstanding something, or your editor tricked you somehow.

Reply:
I was asking about `previousAst`. I don't see it's checked before the foreach loop.
For some context, I suspect that you are expecting directly nested hashtable's, not something like this `@{ abc = @( @{d<tab>} ) }`. However, the code doesn't check if `previewAst` is a hashtable, and thus it's populating `nestedHashtableKeys` even it's not directly nested hashtable's.

Reply:
Ahh okay, my mistake then. I don't remember if I considered this scenario but I don't think it's necessarily wrong. Consider the following scenario:

```
class Level1
{
    [Level2[]] $Prop1
}
class Level2
{
    [string] $Prop2
}
[Level1] @{
    Prop1 = @(
        @{Prop2="Hello1"}
        @{Prop2="Hello2"}
    )
}
```
It would make sense to offer completions for the "Level2" properties when we can infer that the user is building an array of "Level2" objects for the "Level1" type.  
The current implementation isn't perfect though. It doesn't work when the types haven't been loaded yet and `Get-WinEvent -FilterHashtable @{SuppressHashFilter=@( @{<Tab>} )` offers completions even though arrays aren't valid for "SupressHashFilter".
I will have to think about how to solve this.

Reply:
That's a good example. It will be nice if this example can be put in code as comment.

This comment and https://github.com/PowerShell/PowerShell/pull/17660#discussion_r1176890611 below are about the same thing. Maybe think about if we can restrict the checks some more in this while loop.

Reply:
I've updated the check to only allow Asts I expect to see in relevant completion scenarios. I know some people frown on "goto" but it seems like the best solution here. Feel free to correct me if I'm wrong though.